File: flow/script/config/transcribe.ts
/**
 * Transcription service configuration.
 * Fault-tolerant chain: gcp-vertex -> gemini
 */

/** Transcription backend type */
export type TranscriberBackend = "gcp-vertex" | "gemini";

/** Configuration loaded from environment */
export interface TranscribeConfig {
  /** Explicit backend choice; null = auto-failover */
  transcriber: TranscriberBackend | null;

  // GCP Vertex AI
  gcpProjectId: string;
  gcpLocation: string;
  gcpModel: string;
  gcsBucket: string;
  gcsPrefix: string;
  apiVersion: string;
  gcpApiKey: string;

  // Gemini Standard API (fallback)
  geminiStdModel: string;
  googleApiKey: string;
}

function getEnv(key: string, fallback = ""): string {
  return Deno.env.get(key) || fallback;
}

export function loadConfig(): TranscribeConfig {
  const explicit = getEnv("TRANSCRIBER") as TranscriberBackend | "";

  return {
    transcriber: explicit === "gcp-vertex" || explicit === "gemini"
      ? explicit
      : null,

    gcpProjectId: getEnv("GCP_PROJECT_ID", "gen-lang-client-0385617544"),
    gcpLocation: getEnv("GCP_LOCATION", "us-central1"),
    gcpModel: getEnv("GCP_MODEL", "gemini-2.5-flash-lite"),
    gcsBucket: getEnv("GCS_BUCKET", "hermes_brain"),
    gcsPrefix: getEnv("GCS_PREFIX", "audio_jobs/"),
    apiVersion: getEnv("API_VERSION", "v1"),
    gcpApiKey: getEnv("gcp-vertex-key", ""),

    geminiStdModel: getEnv("GEMINI_STD_MODEL", "gemini-2.5-flash-lite"),
    googleApiKey: getEnv("GEMINI_API_KEY", ""),
  };
}

/** Validate that at least one backend has credentials */
export function validateConfig(cfg: TranscribeConfig): void {
  const hasVertex = cfg.gcpApiKey.length > 0;
  const hasGemini = cfg.googleApiKey.length > 0;

  if (!hasVertex && !hasGemini) {
    throw new Error(
      "No transcription credentials found. " +
        "Set gcp-vertex-key or GEMINI_API_KEY environment variable.",
    );
  }
}

File: flow/script/config/transcribe.py
"""
Transcription service configuration.
Fault-tolerant chain: gcp-vertex -> gemini
"""

import os
from typing import Literal, Optional

TranscriberBackend = Literal["gcp-vertex", "gemini"]


def _env(key: str, fallback: str = "") -> str:
    return os.environ.get(key, fallback)


class TranscribeConfig:
    """Configuration loaded from environment."""

    def __init__(self):
        explicit = _env("TRANSCRIBER")
        self.transcriber: Optional[TranscriberBackend] = None
        if explicit in ("gcp-vertex", "gemini"):
            self.transcriber = explicit  # type: ignore[assignment]

        # GCP Vertex AI
        self.gcp_project_id = _env("GCP_PROJECT_ID", "gen-lang-client-0385617544")
        self.gcp_location = _env("GCP_LOCATION", "us-central1")
        self.gcp_model = _env("GCP_MODEL", "gemini-2.5-pro")
        self.gcs_bucket = _env("GCS_BUCKET", "hermes_brain")
        self.gcs_prefix = _env("GCS_PREFIX", "audio_jobs/")
        self.api_version = _env("API_VERSION", "v1")
        self.gcp_api_key = _env("gcp-vertex-key", "")

        # Gemini Standard API (fallback)
        self.gemini_std_model = _env("GEMINI_STD_MODEL", "gemini-2.5-flash-lite")
        self.google_api_key = _env("GEMINI_API_KEY", "")

    def validate(self) -> None:
        """Validate that at least one backend has credentials."""
        if not self.gcp_api_key and not self.google_api_key:
            raise RuntimeError(
                "No transcription credentials found. "
                "Set gcp-vertex-key or GEMINI_API_KEY environment variable."
            )


def load_config() -> TranscribeConfig:
    return TranscribeConfig()
